### PR TITLE
Bug fix/applicationserver stop

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

The goal of this PR is to make `ApplicationServer::beginShutdown()` and `ApplicationServer::stop()` non-overlapping.
Besides that, it refactors the code a bit so that ApplicationServer and ApplicationFeature states are moved to the correct places (classes), things are made private, and that we get rid of the extra `_stopping` flag.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behaviour change can only be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *all Jenkins tests*.

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5092/

### CODE REVIEW

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5100/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5101/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5104/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5106/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5107/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5109/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5119/

File: lib/ApplicationFeatures/ApplicationServer.cpp [hunk 1]
```diff
@@ -733,16 +779,24 @@ void ApplicationServer::unprepare() {
...
+  while (true) {
+    // wait until somebody calls beginShutdown and it finishes
     CONDITION_LOCKER(guard, _shutdownCondition);
+    
+    if (_abortWaiting) {
+      // yippieh!
+      break;
+    }
+
     guard.wait(100000);
```
Comment (goedderz) [inline lib/ApplicationFeatures/ApplicationServer.cpp]: Using the chrono literals would make this slightly more readable and typesafe:
```suggestion
    using namespace std::chrono_literals;
    guard.wait(100000us);
```
or
```suggestion
    using namespace std::chrono_literals;
    guard.wait(100ms);
```

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5125/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5129/

**Pull request merged**